src/jmclient/wallet.py

```diff
@@ -564,21 +567,31 @@ def get_internal_script(self, mixdepth):
 
     @classmethod
     def addr_to_script(cls, addr):
+        """
+        Try not to call this slow method. Instead, call addr_to_path,
+        followed by get_script_from_path, as those are cached.
+        """
```

Comment:
It is faster.
But on reflection there are cases where I don't think we should use a cache.

One of the biggest paranoias in wallet development, is *somehow* giving a user a destination for their coins which is wrong.

It's interesting to consider, just in general: syncing vs 'calculate this thing once so I can broadcast a transaction'. In the latter cases caching for performance is much less important (if at all.. perhaps for big txs). In that situation personally I do *not* want to have the code prefer a lookup over a cheap, but much more trustworthy cryptographic calculation.

I suppose one way to do it is to have callers be able to specify "no-cache" when calling things like `addr_to_script` or `script_to_addr`.

As an example of this, notice how in the current version of the code a call to `script_to_addr` calls `BTCEngine.script_to_address` which not only checks that the script is valid according to bitcoin Script rules (as implemented in `python-bitcointx` but also the cast `CCoinAddress.from_scriptPubKey()` carries out further checks.

You could counter that 'well if the code is written correctly the cache will not be wrong or corrupted etc.' but I think, especially in case we are dealing with persistence to disk, it isn't something I want to fully rely on (to be fair RAM isn't perfect either..)

Reply:
So the TLDR: 'i trust it less if it's coming from disk persistence, not calculation within this program execution' and "no-cache" (however we specify that) really means "take it from the cache but then verify that it's correct and if it isn't please blow up somehow", I guess.

Reply:
Could add options to `wallet-tool.py` to ignore cache and to clear cache?

Reply:
Scope creep. 😑

Reply:
I also don't think we need, or actually even want, to expose this as a decision to users. There's a natural logic to the decision in order to maximize the user's safety, without impacting the performance improvements achieved here.

(There is always an argument 'more options for the user are better' and Joinmarket tends a lot more in that direction than most projects, but I think this would be too far).

And such a thing could be suggested as an addition PR to prevent the scope creep problem, I personally wouldn't be enthusiastic but that's fine.

Reply:
@AdamISZ Could be later, but should be before next release.

Reply:
Another thought: it may be logical to just add this as part of the `--recoversync` option.

Reply:
Agree, adding this to `--recoversync` makes sense.